### PR TITLE
[Feature]: Support CondInst 

### PR DESCRIPTION
**Pull request by YvesXu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add CondInst Head

## Modification

add configs/condinst/condinst_r50_1x8_coco.py
add mmdet/models/dense_head/condinst_head.py
add mmdet/models/detectors/condinst.py

modify mmdet/models/dense_head/__init__.py
modify mmdet/models/detectors/__init__.py

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by jbwang1997** [comment]
Hello @YvesXu, thank your contributions. Please first sign the CLA. 
image
And seems the mmdetection you work on is not the latest version. Please rebase it to the latest version.

File: configs/condinst/condinst_r50_1x8_coco.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    dict(type='Resize', scale=(input_size, input_size), keep_ratio=False),
+    dict(type='RandomFlip', prob=0.5),
+    dict(
+        type='PhotoMetricDistortion',
+        brightness_delta=32,
+        contrast_range=(0.5, 1.5),
+        saturation_range=(0.5, 1.5),
+        hue_delta=18),
+    dict(type='PackDetInputs')
+]
```
Comment (jbwang1997) [inline configs/condinst/condinst_r50_1x8_coco.py:109]: The pipeline seems different from the original implementation in [AdelaiDet](https://github.com/aim-uofa/AdelaiDet/tree/master/configs/CondInst).

File: mmdet/models/dense_heads/condinst_head.py [hunk 2]
```diff
@@ -0,0 +1,1310 @@
...
+        reg_feat = x
+        controller_pred = []
+
+        for cls_layer in self.cls_convs:
+            cls_feat = cls_layer(cls_feat)
+        cls_score = self.conv_cls(cls_feat)
+
+        for reg_layer in self.reg_convs:
+            reg_feat = reg_layer(reg_feat)
+            controller_pred.append[self.controller(reg_feat)]
```
Comment (jbwang1997) [inline mmdet/models/dense_heads/condinst_head.py:81]: Seems have syntax error

File: mmdet/models/dense_heads/condinst_head.py [hunk 3]
```diff
@@ -0,0 +1,1310 @@
...
+    def loss_by_feat(
+        self,
+        cls_scores: List[Tensor],
+        bbox_preds: List[Tensor],
+        centernesses: List[Tensor],
+        controller_pred: List[Tensor],
+        batch_gt_instances: InstanceList,
+        batch_img_metas: List[dict],
+        batch_gt_instances_ignore: OptInstanceList = None
+    ) -> Dict[str, Tensor]:
```
Comment (jbwang1997) [inline mmdet/models/dense_heads/condinst_head.py:114]: Most code is the same with FCOSHead, could we use super to simplify this section?

File: mmdet/models/dense_heads/condinst_head.py [hunk 4]
```diff
@@ -0,0 +1,1310 @@
...
+        # only calculate pos centerness targets, otherwise there may be nan
+        left_right = pos_bbox_targets[:, [0, 2]]
+        top_bottom = pos_bbox_targets[:, [1, 3]]
+        if len(left_right) == 0:
+            centerness_targets = left_right[..., 0]
+        else:
+            centerness_targets = (
+                left_right.min(dim=-1)[0] / left_right.max(dim=-1)[0]) * (
+                    top_bottom.min(dim=-1)[0] / top_bottom.max(dim=-1)[0])
+        return torch.sqrt(centerness_targets)
```
Comment (jbwang1997) [inline mmdet/models/dense_heads/condinst_head.py:383]: Can directly use the function in FCOSHead.

File: mmdet/models/dense_heads/condinst_head.py [hunk 5]
```diff
@@ -0,0 +1,1310 @@
...
+        for in_feature in self.in_features_channels:
+            self.refine.append(kaiming_init(nn.Conv2d(in_feature,self.tower_channel, 3, 1)))
+            # self.refine.append((torch.conv2d(256,self.tower_channel, 3, 1)))
+        self.tower = nn.ModuleList()
+        for i in range(self.num_convs):
+            self.tower.append(kaiming_init(nn.Conv2d(self.tower_channel, self.tower_channel, 3, 1)))
+            # self.refine.append((torch.conv2d(256,self.tower_channel, 3, 1)))
+        self.tower.append(nn.Conv2d(
+            self.tower_channel, max(self.num_outputs, 1), 1
+        ))
```
Comment (jbwang1997) [inline mmdet/models/dense_heads/condinst_head.py:705]: initialization can be unified in _init_weights function.

File: mmdet/models/dense_heads/condinst_head.py [hunk 6]
```diff
@@ -0,0 +1,1310 @@
...
+            0, h * stride, step=stride,
+            dtype=torch.float32, device=device
+        )
+        shift_y, shift_x = torch.meshgrid(shifts_y, shifts_x)
+        shift_x = shift_x.reshape(-1)
+        shift_y = shift_y.reshape(-1)
+        locations = torch.stack((shift_x, shift_y), dim=1) + stride // 2
+        return locations
+
+class SegmentationModule(BaseModule):
```
Comment (jbwang1997) [inline mmdet/models/dense_heads/condinst_head.py:1278]: Seems it is only a conv2d. Is it necessary?

**Comment by jbwang1997** [comment]
Hello @YvesXu, Thanks for your contributions. Please remember to sign the CLA.
image
The lint seems also failed. Please follow the [tutorial](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to fix the lint.

File: configs/condinst/condinst_r50_1x8_coco.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        pad_mask=True),
+    backbone=dict(
+        type='ResNet',
+        depth=50,
+        num_stages=4,
+        out_indices=(0, 1, 2, 3),
+        frozen_stages=1,  # do not freeze stem
+        norm_cfg=dict(type='BN', requires_grad=True),
+        norm_eval=False,  # update the statistics of bn
+        zero_init_residual=False,
```
Comment (hhaAndroid) [inline configs/condinst/condinst_r50_1x8_coco.py:24]: remove

File: configs/condinst/condinst_r50_1x8_coco.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+        mean=img_norm_cfg['mean'],
+        std=img_norm_cfg['std'],
+        bgr_to_rgb=img_norm_cfg['to_rgb'],
+        pad_mask=True),
+    backbone=dict(
+        type='ResNet',
+        depth=50,
+        num_stages=4,
+        out_indices=(0, 1, 2, 3),
+        frozen_stages=1,  # do not freeze stem
```
Comment (hhaAndroid) [inline configs/condinst/condinst_r50_1x8_coco.py:21]: do not freeze stem ?

File: configs/condinst/condinst_r50_1x8_coco.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+            type='FocalLoss',
+            use_sigmoid=True,
+            gamma=2.0,
+            alpha=0.25,
+            loss_weight=1.0),
+        loss_bbox=dict(type='IoULoss', loss_weight=1.0),
+        loss_centerness=dict(
+            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0)),
+    mask_head=dict(
+        type='CondInstDynamicMaskHead',
```
Comment (hhaAndroid) [inline configs/condinst/condinst_r50_1x8_coco.py:52]: DynamicMaskHead

File: configs/condinst/condinst_r50_1x8_coco.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    # training and testing settings
+    train_cfg=dict(
+        assigner=dict(
+            type='MaxIoUAssigner',
+            pos_iou_thr=0.5,
+            neg_iou_thr=0.4,
+            min_pos_iou=0.,
+            ignore_iof_thr=-1,
+            gt_max_assign_all=False),
+        sampler=dict(type='PseudoSampler'),  # YOLACT should use PseudoSampler
```
Comment (hhaAndroid) [inline configs/condinst/condinst_r50_1x8_coco.py:69]: YOLACT?

File: configs/condinst/condinst_r50_1x8_coco.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+]
+
+env_cfg = dict(cudnn_benchmark=True)
+
+# NOTE: `auto_scale_lr` is for automatically scaling LR,
+# USER SHOULD NOT CHANGE ITS VALUES.
+# base_batch_size = (1 GPUs) x (8 samples per GPU)
+auto_scale_lr = dict(base_batch_size=8)
+
+# MODEL:
```
Comment (hhaAndroid) [inline configs/condinst/condinst_r50_1x8_coco.py:161]: del

File: mmdet/models/detectors/condinst.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+from mmdet.registry import MODELS
+from mmdet.utils.typing import ConfigType, OptConfigType, OptMultiConfig
+from .single_stage_instance_seg import SingleStageInstanceSegmentor
+
+
+@MODELS.register_module()
+class CondInst(SingleStageInstanceSegmentor):
+    """Implementation of `CondInst <https://link.springer.com/chapter/10.1007/978-3-030-58452-8_17>`_"""
```
Comment (hhaAndroid) [inline mmdet/models/detectors/condinst.py:10]: arxiv link

File: mmdet/models/dense_heads/condinst_head.py [hunk 13]
```diff
@@ -0,0 +1,1310 @@
...
+        det_bboxes, det_labels, det_coeffs = fast_nms(
+            multi_bboxes, multi_scores, multi_coeffs, cfg.score_thr,
+            cfg.iou_thr, cfg.top_k, cfg.max_per_img)
+        results = InstanceData()
+        results.bboxes = det_bboxes[:, :4]
+        results.scores = det_bboxes[:, -1]
+        results.labels = det_labels
+        results.coeffs = det_coeffs
+        return results
+
```
Comment (BIGWangYuDong) [inline mmdet/models/dense_heads/condinst_head.py:645]: Remember to install pre-commit to check your lint. https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md

File: mmdet/models/dense_heads/condinst_head.py [hunk 14]
```diff
@@ -0,0 +1,1310 @@
...
+            batch_img_metas (list[dict]): Meta information of multiple images.
+            positive_infos (List[:obj:``InstanceData``]): Information of
+                positive samples of each image that are assigned in detection
+                head.
+
+        Returns:
+            dict[str, Tensor]: A dictionary of loss components.
+        """
+        assert positive_infos is not None, \
+            'positive_infos should not be None in `YOLACTProtonet`'
```
Comment (BIGWangYuDong) [inline mmdet/models/dense_heads/condinst_head.py:840]: `YOLACT`?

File: mmdet/models/dense_heads/condinst_head.py [hunk 15]
```diff
@@ -0,0 +1,1310 @@
...
+            if num_pos == 0 or pos_mask_targets is None:
+                loss = mask_preds.sum() * 0.
+                if self.boxinst_enabled:
+                    pass
+            else:
+                #Condinst
+                if self.boxinst_enabled:
+                    pass
+                else:
+                    # gt_inds = pred_instances.gt_inds
```
Comment (BIGWangYuDong) [inline mmdet/models/dense_heads/condinst_head.py:888]: del

**Comment by BIGWangYuDong** [comment]
@YvesXu Thanks for your contribution, please fix the comments and train the CondInst model to make sure the results is same as the original repo

**Pull request closed without merging**